sdv/cag/programmable_constraint.py

```diff
@@ -0,0 +1,196 @@
+
+        Raises:
+            Should raise an error if the metadata is incompatible with the constraint,
+            e.g. an expected table/column is missing or has the wrong sdtype.
+        """
+        return
+
+    def validate_input_data(self, data):
+        """Validates that the data is compatible with the constraint and its parameters.
+
+        Args:
+            data (dict[pd.DataFrame]):
+                The input data to validate.
+
+        Raises:
+            Should raise an error if the input data is incompatible with the constraint,
+            e.g. a column has an unexpected value or incorrect dtype.
+        """
+        return
+
+    def fit(self, data, metadata):
+        """Given data and metadata, the constraint can learn additional info and save for later use.
+
+        Args:
+            metadata (sdv.Metadata):
+                The metadata to fit on.
+            data (dict[pd.DataFrame]):
+                The input data to fit on.
+        """
+        raise NotImplementedError()
+
+    def transform(self, data):
+        """Update the data in order to adhere to the constraint.
+
+        Args:
+            data (dict[pd.DataFrame]):
+                The data to transform.
+
+        Returns:
+            dict[pd.DataFrame]:
+                The transformed data.
+        """
+        raise NotImplementedError()
+
+    def get_updated_metadata(self, metadata):
+        """Returns how the metadata is updated as a result of transforming the data.
+
+        Based on only the metadata and the parameters, the constraint should be able to
+        determine how this metadata would update as a result of transforming the data.
+
+        Args:
+            metadata (sdv.Metadata):
+                The input metadata.
+
+        Returns:
+            sdv.Metadata:
+                The metadata for the transformed data.
+        """
+        raise NotImplementedError()
+
+    def reverse_transform(self, transformed_data):
+        """Given the transformed data, change it back into the original format.
+
+        Args:
+            transformed_data (dict[pd.DataFrame]):
+                The data to reverse transform.
+
+        Returns:
+            dict[pd.DataFrame]:
+                The reverse transformed data.
+        """
+        raise NotImplementedError()
+
+    def is_valid(self, synthetic_data):
+        """Given synthetic data, validate that it matches the constraint.
+
+        Validate that the synthetic data is valid for the constraint by looking at
+        the parameters as well as any learned information
+
+        Args:
+            synthetic_data (dict[pd.DataFrame]):
+                The data to validate.
+
+        Returns:
+            dict[pd.Series]:
+                For each table, a boolean Series indicating whether or not
+                the corresponding row for the table is valid for this constraint.
+        """
+        raise NotImplementedError()
+
+    def fix_data(self, synthetic_data):
```

Comment:
will we rename the `_fix_dayz_data` methods in enterprise to match this?

Reply:
I don't think so. Currently the CAG framework isn't integrated with DayZ so the users have to use the `fix_dayz_data` method, but ideally in the future a user can just add the constraint to DayZ directly and DayZ will call the `fix_dayz_data` method for all the constraints. 

For programmable constraints, all that matters is that the harness's `_fix_dayz_data` method calls the programmable constraint's `fix_data` method.